File: app/node/distort/transform/transformdistortnode.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "transformdistortnode.h"

#include <QGuiApplication>

namespace olive
{

const QString TransformDistortNode::k_parent_input = QStringLiteral("parent_in");
const QString TransformDistortNode::k_texture_input = QStringLiteral("tex_in");
const QString TransformDistortNode::k_autoscale_input =
	QStringLiteral("autoscale_in");
const QString TransformDistortNode::k_interpolation_input =
	QStringLiteral("interpolation_in");

#define super MatrixGenerator

TransformDistortNode::TransformDistortNode()
{
	add_input(k_parent_input, NodeValue::k_matrix);

	add_input(k_autoscale_input, NodeValue::k_combo, 0);

	add_input(k_interpolation_input, NodeValue::k_combo, 2);

	prepend_input(k_texture_input, NodeValue::k_texture,
				 InputFlags(k_input_flag_not_keyframable));

	// Initiate gizmos
	rotation_gizmo_ = add_draggable_gizmo<ScreenGizmo>();
	rotation_gizmo_->add_input(NodeInput(this, k_rotation_input));
	rotation_gizmo_->set_drag_value_behavior(ScreenGizmo::k_absolute);

	poly_gizmo_ = add_draggable_gizmo<PolygonGizmo>();
	poly_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_position_input), 0));
	poly_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_position_input), 1));

	anchor_gizmo_ = add_draggable_gizmo<PointGizmo>();
	anchor_gizmo_->set_shape(PointGizmo::k_anchor_point);
	anchor_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_anchor_input), 0));
	anchor_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_anchor_input), 1));
	anchor_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_position_input), 0));
	anchor_gizmo_->add_input(
		NodeKeyframeTrackReference(NodeInput(this, k_position_input), 1));

	for (int i = 0; i < k_gizmo_scale_count; i++) {
		point_gizmo_[i] = add_draggable_gizmo<PointGizmo>();
		point_gizmo_[i]->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_scale_input), 0));
		point_gizmo_[i]->add_input(
			NodeKeyframeTrackReference(NodeInput(this, k_scale_input), 1));
		point_gizmo_[i]->set_drag_value_behavior(PointGizmo::k_absolute);
	}

	set_flag(k_video_effect);
	set_effect_input(k_texture_input);
}

void TransformDistortNode::retranslate()
{
	super::retranslate();

	set_input_name(k_parent_input, tr("Parent"));
	set_input_name(k_autoscale_input, tr("Auto-Scale"));
	set_input_name(k_texture_input, tr("Texture"));
	set_input_name(k_interpolation_input, tr("Interpolation"));

	set_combo_box_strings(k_autoscale_input,
					   { tr("None"), tr("Fit"), tr("Fill"), tr("Stretch") });
	set_combo_box_strings(k_interpolation_input,
					   { tr("Nearest Neighbor"), tr("Bilinear"),
						 tr("Mipmapped Bilinear") });
}

void TransformDistortNode::value(const NodeValueRow &value,
								 const NodeGlobals &globals,
								 NodeValueTable *table) const
{
	// Generate matrix
	QMatrix4x4 generated_matrix = generate_matrix(
		value, false, false, false, value[k_parent_input].to_matrix());

	// Pop texture
	NodeValue texture_meta = value[k_texture_input];

	TexturePtr job_to_push = nullptr;

	// If we have a texture, generate a matrix and make it happen
	if (TexturePtr texture = texture_meta.to_texture()) {
		// Adjust our matrix by the resolutions involved
		QMatrix4x4 real_matrix = generate_auto_scaled_matrix(
			generated_matrix, value, globals, texture->params());

		if (!real_matrix.isIdentity()) {
			// The matrix will transform things
			ShaderJob job;
			job.insert(QStringLiteral("ove_maintex"), texture_meta);
			job.insert(QStringLiteral("ove_mvpmat"),
					   NodeValue(NodeValue::k_matrix, real_matrix, this));
			job.set_interpolation(QStringLiteral("ove_maintex"),
								 static_cast<Texture::Interpolation>(
									 value[k_interpolation_input].to_int()));

			// Use global resolution rather than texture resolution because this may result in a size change
			job_to_push = Texture::job(globals.vparams(), job);
		}
	}

	table->push(NodeValue::k_matrix, QVariant::fromValue(generated_matrix),
				this);

	if (!job_to_push) {
		// Re-push whatever value we received
		table->push(texture_meta);
	} else {
		table->push(NodeValue::k_texture, job_to_push, this);
	}
}

ShaderCode
TransformDistortNode::get_shader_code(const ShaderRequest &request) const
{
	Q_UNUSED(request);

	// Returns default frag and vert shader
	return ShaderCode();
}

void TransformDistortNode::gizmo_drag_start(const NodeValueRow &row, double x,
										  double y, const Rational &time)
{
	DraggableGizmo *gizmo = static_cast<DraggableGizmo *>(sender());

	if (gizmo == anchor_gizmo_) {
		gizmo_inverted_transform_ =
			generate_matrix(row, true, true, false, row[k_parent_input].to_matrix())
				.toTransform()
				.inverted();

	} else if (is_a_scale_gizmo(gizmo)) {
		// Dragging scale handle
		TexturePtr tex = row[k_texture_input].to_texture();
		if (!tex) {
			return;
		}

		gizmo_scale_uniform_ = row[k_uniform_scale_input].to_bool();
		gizmo_anchor_pt_ = (row[k_anchor_input].to_vec2() +
							gizmo->get_globals().nonsquare_resolution() / 2)
							   .toPointF();

		if (gizmo == point_gizmo_[k_gizmo_scale_top_left] ||
			gizmo == point_gizmo_[k_gizmo_scale_top_right] ||
			gizmo == point_gizmo_[k_gizmo_scale_bottom_left] ||
			gizmo == point_gizmo_[k_gizmo_scale_bottom_right]) {
			gizmo_scale_axes_ = k_gizmo_scale_both;
		} else if (gizmo == point_gizmo_[k_gizmo_scale_center_left] ||
				   gizmo == point_gizmo_[k_gizmo_scale_center_right]) {
			gizmo_scale_axes_ = k_gizmo_scale_x_only;
		} else {
			gizmo_scale_axes_ = k_gizmo_scale_y_only;
		}

		// Store texture size
		VideoParams texture_params = tex->params();
		QVector2D texture_sz(texture_params.square_pixel_width(),
							 texture_params.height());
		gizmo_scale_anchor_ = row[k_anchor_input].to_vec2() + texture_sz / 2;

		if (gizmo == point_gizmo_[k_gizmo_scale_top_right] ||
			gizmo == point_gizmo_[k_gizmo_scale_bottom_right] ||
			gizmo == point_gizmo_[k_gizmo_scale_center_right]) {
			// Right handles, flip X axis
			gizmo_scale_anchor_.setX(texture_sz.x() - gizmo_scale_anchor_.x());
		}

		if (gizmo == point_gizmo_[k_gizmo_scale_bottom_left] ||
			gizmo == point_gizmo_[k_gizmo_scale_bottom_right] ||
			gizmo == point_gizmo_[k_gizmo_scale_bottom_center]) {
			// Bottom handles, flip Y axis
			gizmo_scale_anchor_.setY(texture_sz.y() - gizmo_scale_anchor_.y());
		}

		// Store current matrix
		gizmo_inverted_transform_ =
			generate_matrix(row, true, true, true, row[k_parent_input].to_matrix())
				.toTransform()
				.inverted();

	} else if (gizmo == rotation_gizmo_) {
		gizmo_anchor_pt_ = (row[k_anchor_input].to_vec2() +
							gizmo->get_globals().nonsquare_resolution() / 2)
							   .toPointF();
		gizmo_start_angle_ =
			std::atan2(y - gizmo_anchor_pt_.y(), x - gizmo_anchor_pt_.x());
		gizmo_last_angle_ = gizmo_start_angle_;
		gizmo_last_alt_angle_ =
			std::atan2(x - gizmo_anchor_pt_.x(), y - gizmo_anchor_pt_.y());
		gizmo_rotate_wrap_ = 0;
		gizmo_rotate_last_dir_ = k_direction_none;
	}
}

void TransformDistortNode::gizmo_drag_move(double x, double y,
										 const Qt::KeyboardModifiers &modifiers)
{
	DraggableGizmo *gizmo = static_cast<DraggableGizmo *>(sender());

	if (gizmo == poly_gizmo_) {
		NodeInputDragger &x_drag = gizmo->get_draggers()[0];
		NodeInputDragger &y_drag = gizmo->get_draggers()[1];

		x_drag.drag(x_drag.get_start_value().toDouble() + x);
		y_drag.drag(y_drag.get_start_value().toDouble() + y);

	} else if (gizmo == anchor_gizmo_) {
		NodeInputDragger &x_anchor_drag = gizmo->get_draggers()[0];
		NodeInputDragger &y_anchor_drag = gizmo->get_draggers()[1];
		NodeInputDragger &x_pos_drag = gizmo->get_draggers()[2];
		NodeInputDragger &y_pos_drag = gizmo->get_draggers()[3];

		QPointF inverted_movement(gizmo_inverted_transform_.map(QPointF(x, y)));

		x_anchor_drag.drag(x_anchor_drag.get_start_value().toDouble() +
						   inverted_movement.x());
		y_anchor_drag.drag(y_anchor_drag.get_start_value().toDouble() +
						   inverted_movement.y());
		x_pos_drag.drag(x_pos_drag.get_start_value().toDouble() + x);
		y_pos_drag.drag(y_pos_drag.get_start_value().toDouble() + y);

	} else if (gizmo == rotation_gizmo_) {
		double raw_angle =
			std::atan2(y - gizmo_anchor_pt_.y(), x - gizmo_anchor_pt_.x());
		double alt_angle =
			std::atan2(x - gizmo_anchor_pt_.x(), y - gizmo_anchor_pt_.y());

		double current_angle = raw_angle;

		// Detect rotation wrap around
		RotationDirection this_dir =
			get_direction_from_angles(gizmo_last_angle_, raw_angle);
		RotationDirection alt_dir =
			get_direction_from_angles(gizmo_last_alt_angle_, alt_angle);

		if (gizmo_rotate_last_dir_ != k_direction_none &&
			this_dir != gizmo_rotate_last_dir_) {
			if (alt_dir == gizmo_rotate_last_alt_dir_) {
				if ((raw_angle - gizmo_last_angle_) < 0) {
					gizmo_rotate_wrap_++;
				} else {
					gizmo_rotate_wrap_--;
				}

				this_dir = gizmo_rotate_last_dir_;
				alt_dir = gizmo_rotate_last_alt_dir_;
			}
		}

		gizmo_rotate_last_dir_ = this_dir;
		gizmo_rotate_last_alt_dir_ = alt_dir;
		gizmo_last_angle_ = raw_angle;
		gizmo_last_alt_angle_ = alt_angle;

		current_angle += M_PI * 2 * gizmo_rotate_wrap_;

		// Convert radians to degrees
		double rotation_difference =
			(current_angle - gizmo_start_angle_) * 57.2958;

		NodeInputDragger &d = gizmo->get_draggers()[0];
		d.drag(d.get_start_value().toDouble() + rotation_difference);

	} else if (is_a_scale_gizmo(gizmo)) {
		QPointF mouse_relative =
			gizmo_inverted_transform_.map(QPointF(x, y) - gizmo_anchor_pt_);

		double x_scaled_movement =
			qAbs(mouse_relative.x() / gizmo_scale_anchor_.x());
		double y_scaled_movement =
			qAbs(mouse_relative.y() / gizmo_scale_anchor_.y());

		NodeInputDragger &x_drag = gizmo->get_draggers()[0];
		NodeInputDragger &y_drag = gizmo->get_draggers()[1];

		switch (gizmo_scale_axes_) {
		case k_gizmo_scale_x_only:
			x_drag.drag(x_scaled_movement);
			break;
		case k_gizmo_scale_y_only:
			if (gizmo_scale_uniform_) {
				x_drag.drag(y_scaled_movement);
			} else {
				y_drag.drag(y_scaled_movement);
			}
			break;
		case k_gizmo_scale_both:
			if (gizmo_scale_uniform_) {
				double distance =
					std::hypot(mouse_relative.x(), mouse_relative.y());
				double texture_diag = std::hypot(gizmo_scale_anchor_.x(),
												 gizmo_scale_anchor_.y());

				x_drag.drag(qAbs(distance / texture_diag));
			} else {
				x_drag.drag(x_scaled_movement);
				y_drag.drag(y_scaled_movement);
			}
			break;
		}
	}
}

QMatrix4x4 TransformDistortNode::adjust_matrix_by_resolutions(
	const QMatrix4x4 &mat, const QVector2D &sequence_res,
	const QVector2D &texture_res, const QVector2D &offset,
	AutoScaleType autoscale_type)
{
	// First, create an identity matrix
	QMatrix4x4 adjusted_matrix;

	// Scale it to a square based on the sequence's resolution
	adjusted_matrix.scale(2.0 / sequence_res.x(), 2.0 / sequence_res.y(), 1.0);

	// Apply offset if applicable
	adjusted_matrix.translate(offset.x(), offset.y());

	// Adjust by the matrix we generated earlier
	adjusted_matrix *= mat;

	// Scale back out to texture size (adjusted by pixel aspect)
	adjusted_matrix.scale(texture_res.x() * 0.5, texture_res.y() * 0.5, 1.0);

	// If auto-scale is enabled, fit the texture to the sequence (without cropping)
	if (autoscale_type != k_auto_scale_none) {
		if (autoscale_type == k_auto_scale_stretch) {
			adjusted_matrix.scale(sequence_res.x() / texture_res.x(),
								  sequence_res.y() / texture_res.y(), 1.0);
		} else {
			double footage_real_ar = texture_res.x() / texture_res.y();
			double sequence_real_ar = sequence_res.x() / sequence_res.y();

			double scale_by_x = sequence_res.x() / texture_res.x();
			double scale_by_y = sequence_res.y() / texture_res.y();
			double autoscale_val;

			if ((autoscale_type == k_auto_scale_fit) ==
				(sequence_real_ar > footage_real_ar)) {
				// Scale by height. Either the sequence is wider than the footage or we're using fill and
				// cutting off the sides
				autoscale_val = scale_by_y;
			} else {
				// Scale by width. Either the footage is wider than the sequence or we're using fill and
				// cutting off the top and bottom
				autoscale_val = scale_by_x;
			}

			adjusted_matrix.scale(autoscale_val, autoscale_val, 1.0);
		}
	}

	return adjusted_matrix;
}

void TransformDistortNode::update_gizmo_positions(const NodeValueRow &row,
												const NodeGlobals &globals)
{
	TexturePtr tex = row[k_texture_input].to_texture();
	if (!tex) {
		return;
	}

	// Get the sequence resolution
	const QVector2D &sequence_res = globals.nonsquare_resolution();
	QVector2D sequence_half_res = sequence_res * 0.5;
	QPointF sequence_half_res_pt = sequence_half_res.toPointF();

	// GizmoTraverser just returns the sizes of the textures and no other data
	VideoParams tex_params = tex->params();
	QVector2D tex_sz(tex_params.square_pixel_width(), tex_params.height());
	QVector2D tex_offset = tex_params.offset();

	// Retrieve autoscale value
	AutoScaleType autoscale =
		static_cast<AutoScaleType>(row[k_autoscale_input].to_int());

	// Fold values into a matrix for the rectangle
	QMatrix4x4 rectangle_matrix;
	rectangle_matrix.scale(sequence_half_res.x(), sequence_half_res.y());
	rectangle_matrix *= adjust_matrix_by_resolutions(
		generate_matrix(row, false, false, false, row[k_parent_input].to_matrix()),
		sequence_res, tex_sz, tex_offset, autoscale);

	// Create rect and transform it
	const QVector<QPointF> points = { QPointF(-1, -1), QPointF(1, -1),
									  QPointF(1, 1), QPointF(-1, 1),
									  QPointF(-1, -1) };
	QTransform rectangle_transform = rectangle_matrix.toTransform();
	QPolygonF r = rectangle_transform.map(points);
	r.translate(sequence_half_res_pt);
	poly_gizmo_->set_polygon(r);

	// Draw anchor point
	QMatrix4x4 anchor_matrix;
	anchor_matrix.scale(sequence_half_res.x(), sequence_half_res.y());
	anchor_matrix *= adjust_matrix_by_resolutions(
		generate_matrix(row, true, false, false, row[k_parent_input].to_matrix()),
		sequence_res, tex_sz, tex_offset, autoscale);
	anchor_gizmo_->set_point(anchor_matrix.toTransform().map(QPointF(0, 0)) +
							sequence_half_res_pt);

	// Draw scale handles
	point_gizmo_[k_gizmo_scale_top_left]->set_point(
		create_scale_point(-1, -1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_top_center]->set_point(
		create_scale_point(0, -1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_top_right]->set_point(
		create_scale_point(1, -1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_bottom_left]->set_point(
		create_scale_point(-1, 1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_bottom_center]->set_point(
		create_scale_point(0, 1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_bottom_right]->set_point(
		create_scale_point(1, 1, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_center_left]->set_point(
		create_scale_point(-1, 0, sequence_half_res_pt, rectangle_matrix));
	point_gizmo_[k_gizmo_scale_center_right]->set_point(
		create_scale_point(1, 0, sequence_half_res_pt, rectangle_matrix));

	// Use offsets to make the appearance of values that start in the top left, even though we
	// really anchor around the center
	set_input_property(k_position_input, QStringLiteral("offset"),
					 sequence_half_res + tex_offset);
	set_input_property(k_anchor_input, QStringLiteral("offset"), tex_sz * 0.5);
}

QTransform
TransformDistortNode::gizmo_transformation(const NodeValueRow &row,
										  const NodeGlobals &globals) const
{
	if (TexturePtr texture = row[k_texture_input].to_texture()) {
		//auto m = GenerateMatrix(row, false, false, false, row[kParentInput].toMatrix());
		auto m = generate_matrix(row, false, false, false, QMatrix4x4());
		return generate_auto_scaled_matrix(m, row, globals, texture->params())
			.toTransform();
	}
	return super::gizmo_transformation(row, globals);
}

QPointF TransformDistortNode::create_scale_point(double x, double y,
											   const QPointF &half_res,
											   const QMatrix4x4 &mat)
{
	return mat.map(QPointF(x, y)) + half_res;
}

QMatrix4x4 TransformDistortNode::generate_auto_scaled_matrix(
	const QMatrix4x4 &generated_matrix, const NodeValueRow &value,
	const NodeGlobals &globals, const VideoParams &texture_params) const
{
	const QVector2D &sequence_res = globals.nonsquare_resolution();
	QVector2D texture_res(texture_params.square_pixel_width(),
						  texture_params.height());
	AutoScaleType autoscale =
		static_cast<AutoScaleType>(value[k_autoscale_input].to_int());

	return adjust_matrix_by_resolutions(generated_matrix, sequence_res,
									 texture_res, texture_params.offset(),
									 autoscale);
}

bool TransformDistortNode::is_a_scale_gizmo(NodeGizmo *g) const
{
	for (int i = 0; i < k_gizmo_scale_count; i++) {
		if (point_gizmo_[i] == g) {
			return true;
		}
	}

	return false;
}

TransformDistortNode::RotationDirection
TransformDistortNode::get_direction_from_angles(double last, double current)
{
	return (current > last) ? k_direction_positive : k_direction_negative;
}

}

File: src/node/rust/src/nodes/transformdistortnode.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Transform distort effect (C++
//! `src/node/src/distort/transform/transformdistortnode.{h,cpp}`,
//! `olive::TransformDistortNode`). In C++ this derives from
//! `MatrixGenerator` (`super::matrix`), which provides the
//! `pos_in`/`rot_in`/`scale_in`/`uniform_scale_in`/`anchor_in` inputs
//! and the `generate_matrix` helper.

use crate::factory::NodeMeta;
use crate::node::{Category, Gizmo, NodeBehavior, NodeCore};

/// Parent matrix input id (C++ `k_parent_input`). Type: matrix; no
/// default (identity when unconnected).
pub const PARENT_INPUT: &str = "parent_in";

/// Texture input id (C++ `k_texture_input`). Type: texture; flags:
/// not-keyframable; this is the node's effect input. Note: C++
/// `prepend_input`s it so it appears before the inherited matrix
/// inputs.
pub const TEXTURE_INPUT: &str = "tex_in";

/// Auto-scale combo input id (C++ `k_autoscale_input`). Type: combo;
/// default `0` (`AutoScaleType::None`); combo strings: "None", "Fit",
/// "Fill", "Stretch".
pub const AUTOSCALE_INPUT: &str = "autoscale_in";

/// Interpolation combo input id (C++ `k_interpolation_input`). Type:
/// combo; default `2` (mipmapped bilinear); combo strings: "Nearest
/// Neighbor", "Bilinear", "Mipmapped Bilinear".
pub const INTERPOLATION_INPUT: &str = "interpolation_in";

/// Number of scale point gizmos (C++ `k_gizmo_scale_count` from
/// `node.h`: top-left, top-center, top-right, bottom-left,
/// bottom-center, bottom-right, center-left, center-right).
pub const GIZMO_SCALE_COUNT: usize = 8;

/// Auto-scale mode (C++ `AutoScaleType`); values match the
/// `autoscale_in` combo indices.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum AutoScaleType {
	/// No auto-scaling.
	None = 0,
	/// Fit the texture inside the sequence frame (letterbox).
	Fit = 1,
	/// Fill the sequence frame (crop overflow).
	Fill = 2,
	/// Stretch the texture to the sequence frame ignoring aspect.
	Stretch = 3,
}

/// Rotation direction for wrap-around detection (C++ private enum
/// `RotationDirection`).
#[derive(Clone, Copy, PartialEq, Eq)]
enum RotationDirection {
	/// No direction established yet.
	None,
	/// Clockwise (C++ `k_direction_positive`).
	Positive,
	/// Counter-clockwise (C++ `k_direction_negative`).
	Negative,
}

/// Which axes a scale gizmo drags (C++ private enum `GizmoScaleType`).
#[derive(Clone, Copy, PartialEq, Eq)]
enum GizmoScaleType {
	/// Horizontal center handles (C++ `k_gizmo_scale_x_only`).
	XOnly,
	/// Vertical center handles (C++ `k_gizmo_scale_y_only`).
	YOnly,
	/// Corner handles (C++ `k_gizmo_scale_both`).
	Both,
}

/// Transform distort node. Applies a 2D position/rotation/scale/anchor
/// transform to a texture, equivalent to multiplying by an orthographic
/// matrix.
///
/// The C++ class also keeps a `Matrix4x4 gizmo_inverted_transform_`
/// drag-state member; `Matrix4x4` has no Rust equivalent in this crate
/// yet, so the field is omitted here (it is recomputed at drag start
/// and consumed within the drag).
pub struct TransformDistortNode {
	/// Inherited matrix-generator state (C++ base class
	/// `MatrixGenerator`; provides the transform inputs and
	/// `generate_matrix`).
	pub matrix: super::matrix::MatrixGenerator,
	/// Angle between the drag-start mouse position and the anchor
	/// (C++ `double gizmo_start_angle_`).
	gizmo_start_angle: f64,
	/// Anchor point in screen space captured at drag start (C++
	/// `PointF gizmo_anchor_pt_`).
	gizmo_anchor_pt: (f64, f64),
	/// Whether uniform scale was on at scale-drag start (C++
	/// `bool gizmo_scale_uniform_`).
	gizmo_scale_uniform: bool,
	/// Previous-frame raw rotation angle (C++ `double
	/// gizmo_last_angle_`).
	gizmo_last_angle: f64,
	/// Previous-frame perpendicular rotation angle used to disambiguate
	/// wrap-around (C++ `double gizmo_last_alt_angle_`).
	gizmo_last_alt_angle: f64,
	/// Number of full revolutions accumulated while rotation dragging
	/// (C++ `int gizmo_rotate_wrap_`).
	gizmo_rotate_wrap: i32,
	/// Last raw rotation direction (C++ `gizmo_rotate_last_dir_`).
	gizmo_rotate_last_dir: RotationDirection,
	/// Last perpendicular rotation direction (C++
	/// `gizmo_rotate_last_alt_dir_`).
	gizmo_rotate_last_alt_dir: RotationDirection,
	/// Axes dragged by the current scale gizmo (C++
	/// `gizmo_scale_axes_`).
	gizmo_scale_axes: GizmoScaleType,
	/// Scale reference point captured at drag start (texture-space
	/// anchor, flipped for right/bottom handles; C++ `Vector2D
	/// gizmo_scale_anchor_`).
	gizmo_scale_anchor: (f64, f64),
	/// Scale drag handles in `k_gizmo_scale_*` order (C++
	/// `PointGizmo *point_gizmo_[k_gizmo_scale_count]`; absolute drag
	/// value behavior, bound to both tracks of the inherited
	/// `scale_in`).
	point_gizmo: [Gizmo; GIZMO_SCALE_COUNT],
	/// Anchor point handle (C++ `PointGizmo *anchor_gizmo_`; anchor
	/// shape, drags the inherited `anchor_in` and `pos_in` tracks).
	anchor_gizmo: Gizmo,
	/// Frame outline handle (C++ `PolygonGizmo *poly_gizmo_`; drags the
	/// inherited `pos_in` tracks).
	poly_gizmo: Gizmo,
	/// Rotation ring handle (C++ `ScreenGizmo *rotation_gizmo_`;
	/// absolute drag value behavior, bound to the inherited `rot_in`).
	rotation_gizmo: Gizmo,
}

impl TransformDistortNode {
	/// Static matrix composition (C++
	/// `adjust_matrix_by_resolutions()`): scale to a 2x2 square in
	/// sequence space, apply the texture offset and the generated
	/// matrix, scale back out to texture size, then apply the
	/// auto-scale mode (stretch: per-axis; fit/fill: uniform by width
	/// or height depending on the aspect-ratio comparison).
	///
	/// Also covers the C++ private helpers `generate_auto_scaled_matrix`,
	/// `create_scale_point`, `is_a_scale_gizmo` and
	/// `get_direction_from_angles`.
	fn adjust_matrix_by_resolutions() {
		todo!()
	}
}

impl NodeBehavior for TransformDistortNode {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Transform"
	}

	/// Short menu name (C++ `short_name()`; overrides the
	/// `MatrixGenerator` short name "Ortho" to just return `name()`).
	fn short_name(&self) -> &str {
		"Transform"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.transform"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Distort]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Transform an image in 2D space. Equivalent to multiplying by an orthographic matrix."
	}

	/// Localized input names (C++ `retranslate()`): `parent_in` ->
	/// "Parent", `autoscale_in` -> "Auto-Scale" (combo strings "None",
	/// "Fit", "Fill", "Stretch"), `tex_in` -> "Texture",
	/// `interpolation_in` -> "Interpolation" (combo strings "Nearest
	/// Neighbor", "Bilinear", "Mipmapped Bilinear"), plus the inherited
	/// `MatrixGenerator` input names via its retranslate.
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Evaluate outputs (C++ `value()`): generates the matrix from the
	/// inherited transform inputs (position/rotation/scale/anchor,
	/// folded with `parent_in`) and always pushes it as a `k_matrix`
	/// value. With a texture: builds the auto-scaled real matrix via
	/// `adjust_matrix_by_resolutions`; if it is not identity, renders a
	/// shader job at the GLOBAL video params (the transform may change
	/// the size) binding `ove_maintex` and `ove_mvpmat` with the
	/// interpolation selected by `interpolation_in`, and pushes that;
	/// identity matrix (or no texture) -> pass-through push of the
	/// input texture value.
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Shader code request (C++ `get_shader_code()`): ignores the
	/// request id and returns a default (empty) `ShaderCode` — the
	/// node relies on the renderer's default vertex/fragment shaders,
	/// so this maps to `None`.
	fn shader_code(&self, request: &str) -> Option<String> {
		todo!()
	}

	/// Gizmo transform/positions (C++ `update_gizmo_positions()` and
	/// `gizmo_transformation()`). Positions: with a texture, builds the
	/// rectangle matrix (sequence half-res scale folded with the
	/// auto-scaled generated matrix), maps the unit square through it
	/// for the polygon gizmo, places the anchor gizmo at the mapped
	/// origin, the eight scale handles at the mapped unit-square
	/// corners/edge midpoints, and sets the `offset` input property of
	/// `pos_in` (half sequence res + texture offset) and `anchor_in`
	/// (half texture size). Transformation: with a texture returns the
	/// auto-scaled generated matrix (generated with an identity parent);
	/// without one falls back to the base `MatrixGenerator`
	/// transformation.
	fn gizmo_update(&self, core: &NodeCore, row: &crate::value::NodeValueRow) {
		todo!()
	}

	/// Gizmo drag callbacks (C++ `gizmo_drag_start()` /
	/// `gizmo_drag_move()`). Start: anchor gizmo -> stores the inverted
	/// anchor-aware generated matrix; scale handle -> requires a
	/// texture, records uniform-scale state, screen-space anchor point,
	/// per-handle axes (corners both, left/right center X, top/bottom
	/// center Y), the texture-space scale anchor flipped for
	/// right/bottom handles, and the inverted fully-generated matrix;
	/// rotation gizmo -> records the anchor point, start/last angles
	/// (raw and perpendicular) and resets the wrap counter and
	/// direction. Move: polygon gizmo -> drags the position X/Y tracks
	/// by the mouse delta; anchor gizmo -> drags the anchor tracks by
	/// the inverse-matrix-mapped delta and the position tracks by the
	/// raw delta; rotation gizmo -> accumulates wrap-aware angle
	/// (detecting direction reversal across ±pi using the perpendicular
	/// angle) and drags the rotation track by the degree difference
	/// from the start angle; scale handle -> maps the mouse delta into
	/// inverted-matrix space relative to the anchor, then drags the
	/// scale tracks by the normalized magnitude per the handle's axes,
	/// collapsing to a single uniform value (diagonal ratio for corner
	/// handles) when uniform scale is on.
	fn gizmo_drag(&mut self, core: &mut NodeCore, start: bool, x: f64, y: f64, modifiers: u32) {
		todo!()
	}

	/// Deep copy (C++ `copy()`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `TransformDistortNode::TransformDistortNode()`):
/// adds `parent_in` (matrix), `autoscale_in` (combo, default 0) and
/// `interpolation_in` (combo, default 2), and PREPENDS `tex_in`
/// (texture, not-keyframable) ahead of the inherited `MatrixGenerator`
/// inputs; creates the rotation screen gizmo (absolute drag behavior,
/// bound to `rot_in`), the frame polygon gizmo (bound to `pos_in`), the
/// anchor point gizmo (anchor shape, bound to `anchor_in` and `pos_in`)
/// and the eight scale point gizmos (absolute drag behavior, bound to
/// `scale_in`); sets the video-effect flag and the effect input.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ factory entry for
/// `org.olivevideoeditor.Olive.transform`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.transform",
		name: "Transform",
		categories: &[Category::Distort],
		create,
	});
}
